content/Testing in Django - Basics - useful tools and good practices.md

```diff
@@ -0,0 +1,352 @@
+- `setUp` and `tearDown`: These are functions to be executed before (setUp) and after (tearDown) each unit test. Very useful for fixtures.
+- `setUpClass` and `tearDownClass`: In an analogous way of setUp and tearDown, this functions are executed before (setUpClass) and after (tearDownClass) the whole test case. This means that are executed only once for a test case. Since they're used by Django to make important configurations, if you overwrite these methods in your test case class, don't forget to call the `super` implementation inside the functions.
+- `setUpTestData`: this function can be used to have a class-level atomic block to define data for the whole test case. That means that this function automatically rollbacks the changes in the database after the finalization of all the unit tests in the test case. Mainly used to load data to your test database.
+- Assert methods: Django has the set of [assert methods from unittest](https://docs.python.org/3/library/unittest.html#assert-methods), and has [another ones created for the framework](https://docs.djangoproject.com/en/3.2/topics/testing/tools/#assertions). These methods, such as `assertEqual`, `assertIsNone`, `assertTrue`, etc, can be used in the unit test, to check the different conditions that have to be met. For example, in the unit test of login, you can assert that the response has a status code of 200 because you want to make sure that if the data is sent correctly, then your application has to return the response with that code.  
+
+You can avoid using these functions, or use any combination of them. They're no needed, but they're very useful and will help you a lot.
+
+### Test discovery and databases
+Where do you have to put your tests classes? By default, Django recognizes any file that fullfil the pattern `test*.py` under the current working directory. That is, any file under the directory that starts with `test` and of course has the `.py` extension. Inside it, Django will execute any function in the test case class starting with `test`.  
+When you run your tests, you can pass a parameter indicating the desired pattern, just in case you want to use a different one. With that behavior, the place to put the test cases can be anywhere, but I recommend to divide the tests into the several Django apps that you may have in your project. For example, if you have a Django app called `users` that has the models and functionalities related to the users in your project, you can define in there a `test` folder as a module (adding the __init__.py file) with all the tests that cover that important part: tests for login, for sign-up, for user data, account deletion, etc. Something similar to this:
```

Comment:
```suggestion
When you run your tests, you can pass a parameter indicating the desired pattern just in case you want to use a different one. With that behavior, the place to put the test cases can be anywhere, but I recommend dividing the tests into the several Django apps that you may have in your project. For example, if you have a Django app called `users` that has the models and functionalities related to the users in your project, you can define in there a `test` folder as a module (adding the __init__.py file) with all the tests that cover that important part: tests for login, for sign-up, for user data, account deletion, etc. Something similar to this:
```